### PR TITLE
Add user operation support

### PR DESCRIPTION
**Pull request by matthewwalsh0**
## **Description**

Check the type of the selected account when creating transactions.

- If a standard EOA, add the transaction to the `TransactionController` as normal.
- If a 4337 account, instead pass the transaction to the new `UserOperationController` to create an equivalent user operation.

Specifically:

- Create a `UserOperationController` instance in `MetamaskController`.
- Add listeners to the `UserOperationController` to inject transaction state and trigger associated events.
- Encapsulate all the transaction logic from `MetamaskController` into `lib/transaction/util.ts` with unit tests.
- Add `account-abstraction.ts` selectors to query user operation and paymaster state.
- Update the transaction confirmation UI to display an alert if a user operation is using a paymaster.
  - Further UI changes have not been made at this stage given pending refactor of the associated components.

## **Related issues**

Fixes: [#1736](https://github.com/MetaMask/MetaMask-planning/issues/1736) [#1737](https://github.com/MetaMask/MetaMask-planning/issues/1737)

## **Manual testing steps**

No user operation testing is required at this stage as the first 4337 account snap is currently in development and E2E tests will be added in a subsequent PR.

A basic transaction regression is recommended to verify the logic changes in the `MetaMaskController`.

### **Before**

before

### **After**

after

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained what problem this PR is solving and how it is solved.
- [x] I've linked related issues
- [x] I've included manual testing steps
- [x] I've included screenshots/recordings if applicable
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [x] I’ve properly set the pull request status:
  - [x] In case it's not yet "ready for review", I've set it to "draft".
  - [x] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Comment by matthewwalsh0** [comment]
@metamaskbot update-policies

**Comment by matthewwalsh0** [comment]
@metamaskbot update-policies

**Review by OGPoyraz — DISMISSED** [review]

**Comment by matthewwalsh0** [comment]
@metamaskbot update-policies

File: .yarn/patches/@metamask-polling-controller-npm-4.0.0-52c29c5a78.patch [hunk 1]
```diff
@@ -0,0 +1,261 @@
+diff --git a/dist/AbstractPollingController.d.ts b/dist/AbstractPollingController.d.ts
```
[the remark this answers is not in the document]
Comment (matthewwalsh0) [reply .yarn/patches/@metamask-polling-controller-npm-4.0.0-52c29c5a78.patch]: This was a temporary fix pending update of all the controllers, but that's now been done in the extension so this has been deleted.

File: package.json [hunk 2]
```diff
@@ -295,7 +298,8 @@
     "@metamask/snaps-rpc-methods": "^4.1.0",
     "@metamask/snaps-sdk": "^1.3.2",
     "@metamask/snaps-utils": "^5.1.2",
-    "@metamask/transaction-controller": "^19.0.0",
+    "@metamask/transaction-controller": "npm:@metamask-previews/transaction-controller@19.0.1-preview.353ee83b",
```
[the remark this answers is not in the document]
Comment (matthewwalsh0) [reply package.json]: Correct, but it required the latest `20.0.0`.

File: app/scripts/lib/transaction/util.test.ts [hunk 3]
```diff
@@ -0,0 +1,440 @@
...
+          transactionMeta: TRANSACTION_META_MOCK,
+        });
+
+        await expect(addTransaction(request)).rejects.toThrow('Test Error');
+      });
+
+      it('does not wait for result if waitForSubmit is false', async () => {
+        transactionController.addTransaction.mockResolvedValue({
+          result: new Promise(() => {
+            /* Intentionally not resovled */
```
Comment (vinistevam) [inline app/scripts/lib/transaction/util.test.ts:148]:
```suggestion
            /* Intentionally not resolved */
```

File: app/scripts/lib/transaction/util.test.ts [hunk 4]
```diff
@@ -0,0 +1,440 @@
...
+      });
+
+      it('does not wait for transaction hash promise if waitForSubmit is false', async () => {
+        userOperationController.addUserOperationFromTransaction.mockResolvedValue(
+          {
+            id: TRANSACTION_META_MOCK.id,
+            hash: undefined as never,
+            transactionHash: () =>
+              new Promise(() => {
+                /* Intentionally not resovled */
```
Comment (vinistevam) [inline app/scripts/lib/transaction/util.test.ts:232]:
```suggestion
                /* Intentionally not resolved */
```

File: app/scripts/lib/transaction/util.ts [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+  const {
+    networkClientId,
+    transactionController,
+    transactionOptions,
+    transactionParams,
+    userOperationController,
+  } = request;
+
+  const { maxFeePerGas, maxPriorityFeePerGas } = transactionParams;
+  const { origin, requireApproval, type } = transactionOptions as any;
```
Comment (vinistevam) [inline app/scripts/lib/transaction/util.ts:137]: We might not need to cast it here. Unless we need something down the line.

**Review by vinistevam — DISMISSED** [review]

File: app/scripts/lib/transaction/util.ts (hunk 5 above)
Comment (matthewwalsh0) [reply app/scripts/lib/transaction/util.ts:137]: This is intentional for now as the `SmartContractAccount` at this point is required but I have a PR pending to make it optional and use the snap account by default.

File: app/scripts/lib/transaction/util.test.ts (hunk 4 above)
Comment (matthewwalsh0) [reply app/scripts/lib/transaction/util.test.ts]: Done.

File: app/scripts/lib/transaction/util.test.ts (hunk 3 above)
Comment (matthewwalsh0) [reply app/scripts/lib/transaction/util.test.ts]: Done.

**Review by vinistevam — APPROVED** [review]

**Review by OGPoyraz — APPROVED** [review]

**Pull request merged**